src/main/java/org/sagebionetworks/web/client/widget/ReactComponent.java

```diff
@@ -0,0 +1,243 @@
+package org.sagebionetworks.web.client.widget;
+
+import com.google.gwt.dom.client.DivElement;
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.Element;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.HasClickHandlers;
+import com.google.gwt.event.shared.HandlerRegistration;
+import com.google.gwt.user.client.ui.ComplexPanel;
+import com.google.gwt.user.client.ui.Widget;
+import java.util.ArrayList;
+import java.util.List;
+import jsinterop.base.JsPropertyMap;
+import org.sagebionetworks.web.client.jsinterop.React;
+import org.sagebionetworks.web.client.jsinterop.ReactComponentProps;
+import org.sagebionetworks.web.client.jsinterop.ReactComponentType;
+import org.sagebionetworks.web.client.jsinterop.ReactDOM;
+import org.sagebionetworks.web.client.jsinterop.ReactDOMRoot;
+import org.sagebionetworks.web.client.jsinterop.ReactNode;
+
+/**
+ * Widget that manages the lifecycle of a React component. To use this widget, create a {@link ReactNode} using the
+ * {@link React} API and call {@link #render(ReactNode)} to render the React component.
+ * <p>
+ * This widget also manages appending child elements if the associated React component can contain children. If all
+ * child widgets use this class, then the child {@link ReactNode}s will be cloned and passed as children to the React
+ * component. If any child of this component is a non-ReactComponent widget, then the child widgets will be injected
+ * into the node found using the component's `ref`.
+ * <p>
+ * The root element defaults to a `div`, but can be changed (e.g. to a `span`) using the {@link #ReactComponent(String)} constructor.
+ * <p>
+ * This widget automatically unmounts the ReactComponent (if any) when this container is detached/unloaded.
+ */
+public class ReactComponent<T extends ReactComponentProps>
+  extends ComplexPanel
+  implements HasClickHandlers {
+
+  private ReactDOMRoot root;
+  private ReactComponentType<T> reactComponentType;
+  public T props;
+
+  private ReactNode<?> reactElement;
+
+  public ReactComponent() {
+    this(DivElement.TAG);
+  }
+
+  public ReactComponent(String tag) {
+    setElement(Document.get().createElement(tag));
+  }
+
+  boolean allChildrenAreReactComponents() {
+    boolean allChildrenAreReactComponents = getChildren().size() > 0;
+    for (Widget w : getChildren()) {
+      if (!(w instanceof ReactComponent)) {
+        allChildrenAreReactComponents = false;
+        break;
+      }
+    }
+    return allChildrenAreReactComponents;
+  }
+
+  boolean isRenderedAsReactComponentChild() {
+    return (
+      getParent() instanceof ReactComponent &&
+      ((ReactComponent<?>) getParent()).allChildrenAreReactComponents()
```

Comment:
So if a sibling is not a react component, then isRenderedAsReactComponentChild is false?

Reply:
Correct! The behavior if we mixed React component children and injected widgets would be unusual from a dev perspective (and potentially undefined depending on the component's behavior). I think in most cases, your GWT widgets would end up injected in the DOM after the children. So if you had widget XML like

```xml
<WrapperThatIsAReactComponentWidget>
  <NonReactWidget1 />
  <ReactWidget />
  <NonReactWidget2 />
</WrapperThatIsAReactComponent>
```
 
You would end up with HTML in a different order, like:

```xml
<wrapper>
  <!-- React children of wrapper are rendered here -->
  <reactWidget />
  <!-- non-React widgets are injected afterwards -->
  <nonReactWidget1 />
  <nonReactWidget2 />
</wrapper>
```

That being said, this approach has the disadvantage that child ReactComponent widgets that are injected will retain their wrapper (e.g. `div` or `span`) that they would not otherwise have, which itself can cause layout issues.

I opted for that disadvantage since I think it would generally be easier to work around using CSS. And if we didn't want to use CSS, the alternative is migrating the non-React sibling widgets to become React components.